### ISSUE TITLE
Proposal: Support for building query string collection in UriBuilder

### ISSUE THREAD

**Issue by deepumi** [dotnet/corefx#12421]
## Background

Currently UriBuilder class does not have a constructor / method to build query string parameters values. 

## Proposed API

```
public class URIBuilder : System.UriBuilder
{
   private readonly NameValueCollection _queryParameters;

   public URIBuilder(string schemeName, string hostName, string path) : this(schemeName, hostName, path, new NameValueCollection()) { }
	
   public URIBuilder(string schemeName, string hostName, string path, NameValueCollection queryParameters) : base(schemeName, hostName)
   {
		_queryParameters = queryParameters ?? new NameValueCollection();
		Path = path;
   }

   public NameValueCollection QueryParameters => _queryParameters;

   public URIBuilder AddQueryParameter(string query, string value)
   {
		if (query == null) throw new ArgumentNullException(nameof(query));

		if (value == null) value = string.Empty; //allow empty value otherwise throw null exception

		_queryParameters.Add(query, value);

		return this;
	}

	public override string ToString()
	{
		var query = new StringBuilder();
		foreach (var item in _queryParameters.AllKeys)
		{
			query.Append($"{item}={Uri.EscapeUriString(_queryParameters[item])}&");
		}

		if (query.Length > 0)
		{
			query.Length--; //Trim last character '&'
			Query = query.ToString();
		}
		return base.ToString();
	}
}
```

**Comment by svick** [comment]
public class URIBuilder : System.UriBuilder

Are you actually suggesting to create a class that differs from `UriBuilder` only in capitalization? I don't think that's a good idea.

    public NameValueCollection QueryParameters

`NameValueCollection` only implements non-generic collection interfaces. Because of that, I don't think it should be used in new code.

**Comment by deepumi** [comment]
@svick  No, I don't want to add a new class, I wanted the same functionality (method and constructor) to implement in the existing UriBuilder.

**Comment by deepumi** [comment]
@karelz  Thank you for adding the label. 

Since UriBuilder class belongs to System namespace could you please add **area-System** label as well.

**Comment by karelz** [comment]
The area should (typically) match contract assembly (which is System.Runtime.Extensions for UriBuilder).

**Comment by deepumi** [comment]
@karelz Thank you got it.

**Comment by deepumi** [comment]
This API needs to be re-designed, Currently there is no unified API in dotnet BCL world to build a URI with QueryString and Path. (Windows, Web and Mobile got a different way to build Query string and path).

**Comment by karelz** [comment]
Do you mean that your original proposal needs to be updated?

**Comment by deepumi** [comment]
@karelz I meant to say the existing System.UriBuilder

**Comment by deepumi** [comment]
@karelz Any idea when this PR will be reviewed?

**Comment by karelz** [comment]
The API needs to be first reviewed by area experts - @AlexGhiondea @joperezr. They are going through the area now. I am not sure if they will time to finish it this year though.

After that we will need CoreFX API review.

My feedback:
- We will need to change the name. Differentiating between `URIBuilder` and `UriBuilder` is really bad IMO. What other alternatives did you consider?
- How often is this kind of functionality hand-written by apps? (ideally some links to popular GitHub projects, or links to StackOverflow / blog post samples)
  - In general it looks useful, but it is always good to collect more data.
- Usage syntax is mixing up fluent API design and modifications API design, that's usually frowned upon (I tried to suggest something similar recently) - we should choose one (IMO the modifications style). If you choose fluent, we should not use `Builder` in the type name.
- Did you consider integrating the API into `UriBuilder`? Why would we need a new type? (easier implementation is not a good enough reason)
- What about adding just `UriQueryStringBuilder` maybe as inner type `QueryBuilder`?

**Comment by deepumi** [comment]
@karelz Thank you for your feedback.
@AlexGhiondea @joperezr 

> Did you consider integrating the API into UriBuilder? Why would we need a new type? (easier implementation is not a good enough reason)

Yes, indeed. I have mentioned this in my proposal. I would like to add this into System.UriBuilder class itself, so that we don't need to add a new type.

> How often is this kind of functionality hand-written by apps? (ideally some links to popular GitHub projects, or links to StackOverflow / blog post samples)

Well, now a days every single app does some sort of external API calls and currently there is no unified API in BCL world to build a URI with QueryString and Path. (Windows, Web and Mobile got a different way to build Query string and path). Ex: In Asp.Net world we have a type name HttpUtility.ParseQueryString (System.Web).
Also, [please refer #726](https://github.com/aspnet/HttpAbstractions/issues/726)

Based on your feedback I have integrated my proposal to UriBuilder class.

```
namespace System
{
    public class UriBuilder
    {
        private readonly NameValueCollection _queryParameters;

        public UriBuilder(string schemeName, string hostName, string path) : this(schemeName, hostName, path, new NameValueCollection()) { }

        public UriBuilder(string schemeName, string hostName, string path, NameValueCollection queryParameters) : this(schemeName, hostName)
        {
            _queryParameters = queryParameters ?? new NameValueCollection();
            Path = path;
        }

        public UriBuilder AddQueryParameter(string key, string value)
        {
            if (key == null) throw new ArgumentNullException(nameof(key));

            _queryParameters.Add(key, value ?? string.Empty);

            return this;
        }

        public NameValueCollection QueryParameters => _queryParameters;

        private void SetQueryParametersFromUri()
        {
            var query = new StringBuilder();
            foreach (var item in _queryParameters.AllKeys)
            {
                query.Append($"{item}={Uri.EscapeUriString(_queryParameters[item])}&");
            }

            if (query.Length > 0)
            {
                query.Length--;
                Query = query.ToString();
            }
        }

        public override string ToString()
        {
            SetQueryParametersFromUri();
            //proceed with rest of the logic here
            //https://referencesource.microsoft.com/#System/net/System/uribuilder.cs,406
        }
    }
}
```

**Comment by AlexGhiondea** [comment]
@deepumi thanks for this proposal.

As @svick mentioned, for new APIs we are trying to move away from using non-generic types like `NameValueCollection`.

As @karelz mentioned, we are trying to stick to a single way of using our APIs. In this case, introducing the fluent APIs might not be something we want to do.

@CIPop as he might have a different view on this.

**Comment by deepumi** [comment]
@AlexGhiondea Thanks you for reviewing it. 

I also added a gist to have IDictionary and IEnumerable version for the above code. Please check here
[https://gist.github.com/deepumi/d6634e38b98e4552ca4dda19bbff0359](https://gist.github.com/deepumi/d6634e38b98e4552ca4dda19bbff0359)

**Comment by karelz** [comment]
@deepumi can you please post just the API shape. Implementation is secondary. Please think through all usage patterns and how it interacts with other APIs (e.g. setting QueryString vs. just adding one pair via your API).

**Comment by deepumi** [comment]
@karelz sure, thank you.

   ```c#
    public class UriQueryStringBuilder : UriBuilder
    {
        public UriQueryStringBuilder(string schemeName, string hostName, string path) : this(schemeName, hostName, path, null) { }

        public UriQueryStringBuilder(string schemeName, string hostName, string path, IEnumerable<KeyValuePair<string, string>> queryParameters) : this(schemeName, hostName,path,null) {}
        
        public UriQueryStringBuilder(string schemeName, string hostName, string path, IDictionary<string, string> queryParameters) : base(schemeName, hostName) {}

        public IDictionary<string, string> QueryParameters {get;}

        public UriQueryStringBuilder AddQueryParameter(string key, string value) {}

        public override string ToString() {}
    }
```

**Usage**

```c#
   //1. Add query string
    var uri = new UriQueryStringBuilder("https", "api.com", "v1/videos/somevideo")
                      .AddQueryParameter("q", "keword")
                      .AddQueryParameter("s", null);

    var result = uri.ToString(); //https://api.com/v1/videos/somevideo?q=keyword&s=
   
    //2. Build query string collection and add query string
     var dict = new Dictionary<string,string>();
     dict["somthing"] = "value";
     dict["query"] = "another value";

     var uri1 = new UriQueryStringBuilder("https", "api.com", "v1/videos/somevideo", dict)
                                  .AddQueryParameter("q", "keword")
                                  .AddQueryParameter("s", null);
     var result1 = uri1.ToString(); //https://api.com/v1/videos/somevideo?something=value&query=another%20value&q=keyword&s=
  
     //3. Build query string collection only.
     var dict1 = new Dictionary<string, string>(); 
     dict1["somthing"] = "value";
     dict1["query"] = "another value";

     var uri2 = new UriQueryStringBuilder("https", "api.com", "v1/videos/somevideo", dict1);
     var result2 = uri2.ToString(); //https://api.com/v1/videos/somevideo?something=value&query=another%20value
```

**Comment by karelz** [comment]
Do you plan to address the feedback on fluent API model?
What about [Query setter](https://apisof.net/catalog/System.UriBuilder.Query) interaction?

**Comment by deepumi** [comment]
@karelz @AlexGhiondea apologize for the delay in getting back to you.

> Do you plan to address the feedback on fluent API model?

Yes, modified the API model above and usage version as well. 

>What about Query setter interaction?

Query setter interaction goes inside ToString() method like [this](https://gist.github.com/deepumi/d6634e38b98e4552ca4dda19bbff0359#file-uriquerystringbuilder-cs-L45)

**Comment by AlexGhiondea** [comment]
@deepumi can you tell me more about how the `IEnumerable<KeyValuePair<string, string>>` overload is different than the `IDictionary` overload? Do they have different usage patterns?

While this is a somewhat useful addition, I am not sure if it is worth the overhead of adding a new type to the framework. 

@CIPop, @terrajobst what do you think?

**Comment by karelz** [comment]
I don't understand the 2 overloads either. @deepumi we need good real-world usage sample links (code patterns which do something similar).

@AlexGhiondea adding a type would be overkill, however latest proposal doesn't do that - it just modifies current `UriBuilder`. I updated the spec on the top.

**Comment by AlexGhiondea** [comment]
@karelz updating the existing `UriBuilder` would be preferable. I would still like to see real world usage patterns.

(also, I am not seeing the updated proposal that uses just `UriBuilder`)

**Comment by karelz** [comment]
Yeah, I missed that one, sorry. There is proposal in the middle which integrates it into UriBuilder.
@deepumi please announce more carefully all changes you're making to the proposal. It is extremely hard to track otherwise.
I will rewrite the top post to `UriBuilder` modification.

**Comment by deepumi** [comment]
@karelz thank you.

**Comment by CIPop** [comment]
> @CIPop, @terrajobst what do you think?

I will assume the request is to add functionality to the existing UriBuilder class (not a new class).

I'm not yet convinced: there are quite a few libraries that already do this for REST APIs. It would be an interesting addition if the REST implementations could reuse this implementation.

@davidfowl @danroth27 would ASP.Net benefit from such API? If so, please review the API shape above.

**Comment by deepumi** [comment]
@CIPop yes, you are correct, I wanted to add the missing features to the existing UriBuilder class. 

> I'm not yet convinced: there are quite a few libraries that already do this for REST APIs. It would be an interesting addition if the REST implementations could reuse this implementation.

The reason I would like to have this in BCL side because the same API (UriBuilder) can reuse in windows, web and mobile frameworks. Currently building dynamic URL in BCL side I would end up using LINQ and other string operations.

https://api.careeronestop.org/api-explorer/home/index/JobSearch_GetTopLocationsByKeywordAndOnetCode
ex :      
```c#
var qs = new Dictionary<string, string>();
qs["searchId"] = "value";
qs["companyName"] = "value";

var uri = new UriBuilder(Uri.UriSchemeHttps, "api.careeronestop.org")
{
    Path = "/v1/jobsearch/{userId}/{keyword}/{location}/{radius}/{days}",
    Query = string.Join("&", qs.Keys.Select(key => key + "=" + Uri.EscapeUriString(qs[key])))
   //instead of the above linq code it would be great to have a additional parameter in the UriBuilder constructor to accept IDictionary
};
```

**Comment by danroth27** [comment]
In ASP.NET Core we have our own QueryBuilder type: https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.aspnetcore.http.extensions.querybuilder. It's available as a NuGet package: https://www.nuget.org/packages/Microsoft.AspNetCore.Http.Extensions.

**Comment by KLuuKer** [comment]
@danroth27 this issue was created here as per https://github.com/aspnet/HttpAbstractions/issues/726

It would be better to have a more unified class(maybe 2?) that handels all uri\querystring building\parsing.

Also for easier discoverability because for me it also took allot of time (googling in frustration) to find out QueryBuilder existed for a long time while I needed it.

**Comment by thomaslevesque** [comment]
In fact, this functionality already exists in an unexpected place: `HttpUtility.ParseQueryString()` returns a specialized implementation of `NameValueCollection` (private nested class `HttpUtility.HttpQSCollection`), which overrides `ToString()` to correctly format the names and values as a query string.

So if you need a query string builder, you can do this:

```csharp
var builder = HttpUtility.ParseQueryString("");
builder.Add("foo", "bar");
builder.Add("id", "42");
var queryString = builder.ToString();
```

However, this relies on undocumented behavior, so it's probably not a safe approach...
